### DISCUSSION TITLE
Trouble with Type Inference for useInfiniteQuery in TanStack Query v5

### DISCUSSION THREAD

**Discussion by JuhwanLeeKR** [TanStack/query#7075] [Q&A] [answered]
Hello everyone,

I've encountered an issue with type inference when using useInfiniteQuery from TanStack Query v5 in a TypeScript project. Despite explicitly typing the return of my custom hook, the TypeScript compiler does not infer the expected UseInfiniteQueryResult type. Here's a simplified version of my hook:

```tsx
export const useReviewBoardReadListInfiniteQuery = (
  params?: ReviewBoardParams,
  config?: UseInfiniteQueryOptions<ServerPaginationData<ReviewBoardResponse[]>>
): UseInfiniteQueryResult<ServerPaginationData<ReviewBoardResponse[]>> => {
  const fetchFn = async ({
    pageParam = 0,
  }): Promise<ServerPaginationData<ReviewBoardResponse[]>> => {
    // Fetch logic...
  };

  return useInfiniteQuery({
    queryKey: ['reviewBoardReadListInfinite', params],
    queryFn: ({ pageParam }) => fetchFn({ pageParam }),
    initialPageParam: 0,
    // Additional configuration...
    ...config,
  });
};
```

When using this hook in a component, I intended for the data property to be inferred as ServerPaginationData<ReviewBoardResponse[]> | undefined, but instead, TypeScript infers it as UseInfiniteQueryResult<ServerPaginationData<ReviewBoardResponse[]>, Error>, which seems to miss aligning with the explicit return type I provided.

Here's how I use the hook:

```tsx
const { data, fetchNextPage, hasNextPage, isError, isFetchingNextPage } = useReviewBoardReadListInfiniteQuery({
  limit: 15,
  isPublished: true,
});
```

And the error I'm seeing is related to accessing data.pages, where TypeScript doesn't recognize pages as part of data.

I've checked my types and configurations against the documentation but can't seem to pinpoint the issue. Has anyone faced a similar problem or can offer insight into what I might be doing wrong? Any suggestions on how to ensure the correct type inference for useInfiniteQuery would be greatly appreciated.

- TypeScript is showing an error, yet in reality, the code functions correctly.

![Screenshot 2024-03-11 at 14 19 48](https://github.com/TanStack/query/assets/87694150/65d8e1dd-b6d9-45ae-9b89-2d9a09168bd3)
![Screenshot 2024-03-11 at 14 19 57](https://github.com/TanStack/query/assets/87694150/a55a4060-6d65-4943-9367-7466b0829855)

Thank you in advance for your help!

**Comment by JuhwanLeeKR** [comment]
The following code does work, but it seems like I might be using it incorrectly. I would appreciate any help:

```tsx
...
const { data, fetchNextPage, hasNextPage, isError, isFetchingNextPage } = useReviewBoardReadListInfiniteQuery({
    limit: 15,
    isPublished: true,
  })

  const reviews = (data as unknown as InfiniteData<ServerPaginationData<ReviewBoardResponse[]>>)?.pages
    .map((page) => page?.items)
    .flat()
...

**Comment by TkDodo** [comment]
show a typescript playground please

**Reply by JuhwanLeeKR** [reply]
This is my code that I simply created with StackBlitz. 🫡
https://stackblitz.com/edit/stackblitz-starters-lavkw3?file=app%2Fpage.tsx

```js
{
...
  "dependencies": {
    ...
    "@tanstack/react-query": "5.25",
    "next": "14.1",
    "typescript": "^5",
    "@types/node": "18.17"
    ...
  }
}
```

**Reply by TkDodo** [reply, marked as answer]
everything is easier if you don't spread `...config`. You'd need to set all type parameters on `UseInfiniteQueryOptions`

[playground](https://www.typescriptlang.org/play?target=6#code/JYWwDg9gTgLgBAbwFBzgVQM4FMCSA7AM2D2BiwEUBXLKATwHkwZgI8MAaFdbfIksqjVoAlLBkoAbGJ1SUehYqQrU6MuL0VkAIgEMYOzgF84BKBBBwA5AAF9bfQGMA1gHooWHQ5gBaAI4raSwBuJCRiMigCTyw4UQA3YCwAdwAhCB0oABMABQydEAxELmAMbMoAIwkSgAssTIB+AC44cogICQ88ENQqkFImuDxKEHKabrgIAgJsGAGhkbGkQ1CsAA9IWDhwmiiHGIBlGjiaXIBzYj0WPF19AB4AFQA+ItQlAub78YcISjwYZvmoygIWWSDWG3g20i0ViWASyTSGUyogwkDYMWQr0yAOGQPGzBgHWaGBgUGIp3GJTKlRqdWarXanUpGAAcsl6W0Ojoulw5DQADIQBw6CQAMUkEhZ+SwxNJ5PGfKgeGlsrJeApSxW62g8G+9jgfPiiVS6Syoh0mX5JRgGn4yiEcAAvHAABRcMB5AoDI0I005T0cLh6oinAaYXAKO2COiMZisDC3Q5QY5QM4XOPXPQ6W4+k1IlFo7AAbQAuo9HkgAJRO56YuB6kmIOC9UhOuAARgADHBjM6PVB8oV6vVEIYQkH4-ACFgYA5qqK8G2dBhaHgHK6EHAPacsLkBxZDM1N9vd56cQsoD3K81smY+thE0cTjpzsqMzds7nEWaxIWsKXyxrF5UDgFwXDgDBzBiCQIHOBwuFQBt4EoKAJDbSxggQ+tJzgdxCmdF0dCSHRW2nWdqhdFCJEratlzgblaHGVB3BgFDFzw8Yxy4Fi2INeQ+CUaNaBdOtUH8IQAGksFoZoi0sdx4TzH8LStElbSUSx2C3AMSzUMSAgXZoRO0nc93yK8ayw1AyLnBdjJPMyQCMhzPXowpARoathzgbtDErPStn4YARTOU992aTsAp3GA2VWGBQscoyJGXeKXywLSRQkUKMGrR1aysrYCFdZKSVC+oADo3kKAAyarmxS0LKrIApyo6dUYGqOBbmbUBSGrHilQNPBMiwPg6iYkCBsXTLstarB2uqCbDACsDyrW4NgFONQ-JBcdwR1OARqiSQp1+LwrjgAAJKCXWrOskKbTIsx7NtDThY1v2RDxLWtdSBACESsJbf4OwAVgCqkKiqDBamxOBSWobbK3HVAwMGCB4doMAxAcMkmDgGgzCgCd9QU418MOrMKpPDBypAHQwBdF0T1y54TwqqrK3KghkpgW64G80txjRkCAD16iQIA)

**Reply by JuhwanLeeKR** [reply]
Thank you It was a great help 👍
